src/strings_internal.h

```diff
@@ -0,0 +1,123 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file strings_interal.h Types and functions related to the internal workings of formatting OpenTTD's strings. */
+
+#ifndef STRINGS_INTERNAL_H
+#define STRINGS_INTERNAL_H
+
+/**
+ * Equivalent to the std::back_insert_iterator in function, with some
+ * convenience helpers for string concatenation.
+ *
+ * The formatter is currently backed by an external char buffer, and has some
+ * extra functions to ease the migration from char buffers to std::string.
+ */
+class StringBuilder {
+	char *current; ///< The current location to add strings
+	const char *last; ///< The last element of the buffer.
+
+public:
+	/* Required type for this to be an output_iterator; mimicks std::back_insert_iterator. */
+	using value_type = void;
+	using difference_type = void;
+	using iterator_category = std::output_iterator_tag;
+	using pointer = void;
+	using reference = void;
+
+	/**
+	 * Create the builder of an external buffer.
+	 * @param start The start location to write to.
+	 * @param last  The last location to write to.
+	 */
+	StringBuilder(char *start, const char *last) : current(start), last(last) {}
+
+	/* Required operators for this to be an output_iterator; mimicks std::back_insert_iterator. */
+	StringBuilder &operator++() { return *this; }
+	StringBuilder operator++(int) { return *this; }
```

Comment:
Shouldn't there be some ++ on the right side?

Reply:
You return `*this`. So if I do `StringBuilder a; a++; a++; a++`. I get 3 times the same pointer back? Or am I missing something? I miss the actual changing of anything in or with `*this` :)

Reply:
In short: `std::back_insert_iterator` does it this way.
Longer explanations: https://en.cppreference.com/w/cpp/iterator/back_insert_iterator/operator%2B%2B. The functions need to exist for it to be considered an `output_iterator`, but like the `std::back_insert_iterator` it does not do anything as the assignments itself already move the iterator. And it needs to be an `output_iterator` for `fmt::format_to` to function.

Reply:
Please add this tiny piece of information to the comment. You now only note it mimics (misspelled btw) the iterator, but maybe add: `mimics std::back_insert_iterator, which do nothing`. (well, my phrasing sucks, but okay :D)

And tnx for the explanation :)